File: Java/FileUtils.java
package com.example.xemu;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.util.Log;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

public class FileUtils {

    // andstation3 one ps3 emulator 🔥
    public static String getFileName(Context context, Uri uri) {
        if (context == null || uri == null) {
            return null; 
        }
        
        String fileName = null;
        Cursor cursor = null;

        try {
            if (uri.getScheme().equals("content")) {
                cursor = context.getContentResolver().query(uri, null, null, null, null);
                if (cursor != null && cursor.moveToFirst()) {
                    fileName = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
                }
            } else if (uri.getScheme().equals("file")) {
                fileName = new File(uri.getPath()).getName();
            }
        } catch (Exception e) {
            Log.e("FileUtils", "Error getting file name", e);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return fileName;
    }

    // Xbox original 
    public static boolean isXboxOriginalISO(File isoFile) {
        if (isoFile == null || !isoFile.exists()) {
            return false;
        }

        try (FileInputStream fis = new FileInputStream(isoFile)) {
            byte[] buffer = new byte[2048]; 
            int bytesRead = fis.read(buffer, 0, buffer.length);

            if (bytesRead > 0) {
                String header = new String(buffer, 0, bytesRead, StandardCharsets.US_ASCII);
                if (header.contains("MICROSOFT*XBOX*MEDIA")) {
                    return true; 
                }
            }
        } catch (IOException e) {
            Log.e("FileUtils", "Error reading ISO file", e);
        }
        return false;
    }

    public static String getPathFromUri(Context context, Uri uri) {
        if (context == null || uri == null) {
            return null; 
        }
        
        String path = null;
        Cursor cursor = null;

        try {
            if (uri.getScheme().equals("content")) {
                cursor = context.getContentResolver().query(uri, null, null, null, null);
                if (cursor != null && cursor.moveToFirst()) {
                    path = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
                }
            } else if (uri.getScheme().equals("file")) {
                path = uri.getPath();
            }
        } catch (Exception e) {
            Log.e("FileUtils", "Error getting file path", e);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return path;
    }
}

File: app/src/Main/java/net/Xanite/ui/utils/FileUtils.kt
package com.xanite.utils

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import java.io.File
import java.io.InputStream
import java.io.OutputStream

object FileUtils {

    fun getGamesDirectory(context: Context, folderName: String): File {
        return File(context.getExternalFilesDir(null), folderName).apply {
            if (!exists()) {
                mkdirs()
            }
        }
    }

    fun getFileNameFromUri(context: Context, uri: Uri): String? {
        return when (uri.scheme) {
            "content" -> getContentFileName(context, uri)
            "file" -> uri.lastPathSegment?.let { File(it).name }
            else -> null
        }
    }

    fun copyStreamToFile(inputStream: InputStream, destFile: File): Boolean {
        return try {
            inputStream.use { input ->
                destFile.outputStream().use { output ->
                    input.copyTo(output)
                    true
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    private fun getContentFileName(context: Context, uri: Uri): String? {
        return context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                cursor.getString(cursor.getColumnIndexOrThrow(OpenableColumns.DISPLAY_NAME))
            } else {
                null
            }
        }
    }

    fun copyFile(sourceFile: File, destFile: File): Boolean {
        return try {
            sourceFile.inputStream().use { input ->
                destFile.outputStream().use { output ->
                    input.copyTo(output)
                    true
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
}